service/task.go

```diff
@@ -840,67 +791,238 @@ func (uis *UIServer) taskModify(w http.ResponseWriter, r *http.Request) {
+	taskExec, err := strconv.Atoi(taskExecutionsAsString)
+	grip.Warning(err)
 	var (
-		testLog  *model.TestLog
-		err      error
-		taskExec int
+		logReader io.ReadCloser
+		testLog   *model.TestLog
 	)
 
-	if logId != "" { // direct link to a log document by its ID
-		testLog, err = model.FindOneTestLogById(logId)
-		if err != nil {
-			uis.LoggedError(w, r, http.StatusInternalServerError, err)
+	// Check buildlogger logs first.
+	opts := apimodels.GetBuildloggerLogsOptions{
+		BaseURL:       uis.Settings.Cedar.BaseURL,
+		TaskID:        taskID,
+		TestName:      testName,
+		Execution:     taskExec,
+		PrintPriority: !raw,
+	}
+	logReader, err = apimodels.GetBuildloggerLogs(r.Context(), opts)
+	if err == nil {
+		defer func() {
+			grip.Warning(message.WrapError(logReader.Close(), message.Fields{
+				"task_id":   taskID,
+				"test_name": testName,
+				"message":   "failed to close buildlogger log ReadCloser",
+			}))
+		}()
+		data.Buildlogger = make(chan apimodels.LogMessage, 1024)
+		go apimodels.ReadBuildloggerToChan(r.Context(), taskID, logReader, data.Buildlogger)
+	} else {
+		grip.Error(message.WrapError(err, message.Fields{
+			"task_id":   taskID,
+			"test_name": testName,
+			"message":   "problem getting buildlogger logs",
+		}))
+	}
+
+	// If buildlogger fails, fall back to db.
+	if logReader == nil {
+		if logId != "" { // direct link to a log document by its ID
+			testLog, err = model.FindOneTestLogById(logId)
+			if err != nil {
+				uis.LoggedError(w, r, http.StatusInternalServerError, err)
+				return
+			}
+		} else {
+			testLog, err = model.FindOneTestLog(testName, taskID, taskExec)
+			if err != nil {
+				uis.LoggedError(w, r, http.StatusInternalServerError, err)
+				return
+			}
+		}
+
+		if testLog == nil {
+			http.Error(w, "not found", http.StatusNotFound)
 			return
 		}
+
+		ctx, cancel := context.WithCancel(r.Context())
+		defer cancel()
+
+		data.Data = make(chan apimodels.LogMessage)
+		go func() {
+			defer close(data.Data)
+			for _, line := range testLog.Lines {
+				if ctx.Err() != nil {
+					return
+				}
+				data.Data <- apimodels.LogMessage{
+					Type:     apimodels.TaskLogPrefix,
+					Severity: apimodels.LogInfoPrefix,
+					Version:  evergreen.LogmessageCurrentVersion,
+					Message:  line,
+				}
+			}
+		}()
+	}
+
+	template := "task_log.html"
+	if raw {
+		template = "task_log_raw.html"
+		uis.renderText.Stream(w, http.StatusOK, data, "base", template)
 	} else {
-		taskID := vars["task_id"]
-		testName := vars["test_name"]
-		taskExecutionsAsString := vars["task_execution"]
-		taskExec, err = strconv.Atoi(taskExecutionsAsString)
-		if err != nil {
-			http.Error(w, "task execution num must be an int", http.StatusBadRequest)
-			return
+		uis.render.WriteResponse(w, http.StatusOK, data, "base", template)
+	}
+}
+
+func (uis *UIServer) getTestResults(w http.ResponseWriter, r *http.Request, projCtx projectContext, uiTask *uiTaskData, execution int) {
```

Comment:
Can do this by without using `uiTask` as an in-out parameter (e.g. just return a results struct from `getTestResults()`)?

Reply:
I still have this question.

Reply:
There are other fields in the `uiTask` struct that are changed here as well, seems cleaner to do it this way. Do you have a strong opinion about this?